File: python/Railfence.py
def rail_fence_encrypt(text, rails):
    """Encrypts text using Rail Fence Cipher."""
    # Create a list of strings for each rail
    fence = [[] for _ in range(rails)]
    rail = 0
    direction = 1 # 1 = down, -1 = up
    
    for char in text:
        fence[rail].append(char)
        rail += direction
        
        # Bounce when hitting top or bottom rail
        if rail == rails:
            rail = rails - 2
            direction = -1
        elif rail < 0:
            rail = 1
            direction = 1
            
    # Join all rows to get ciphertext
    return "".join("".join(row) for row in fence)

def rail_fence_decrypt(text, rails):
    """Decrypts text using Rail Fence Cipher."""
    # 1. Reconstruct the zigzag pattern with placeholders
    fence = [['\n' for _ in range(len(text))] for _ in range(rails)]
    rail = 0
    direction = 1
    
    # Mark the spots with '*'
    for i in range(len(text)):
        fence[rail][i] = '*'
        rail += direction
        
        if rail == rails:
            rail = rails - 2
            direction = -1
        elif rail < 0:
            rail = 1
            direction = 1
            
    # 2. Fill the fence with the ciphertext
    index = 0
    for r in range(rails):
        for c in range(len(text)):
            if fence[r][c] == '*' and index < len(text):
                fence[r][c] = text[index]
                index += 1
                
    # 3. Read the matrix in zigzag to get plaintext
    result = []
    rail = 0
    direction = 1
    for i in range(len(text)):
        result.append(fence[rail][i])
        rail += direction
        
        if rail == rails:
            rail = rails - 2
            direction = -1
        elif rail < 0:
            rail = 1
            direction = 1
            
    return "".join(result)

# --- Main Execution ---
if __name__ == "__main__":
    text = "HELLO WORLD"
    rails = 2 # Depth
    
    encrypted = rail_fence_encrypt(text, rails)
    decrypted = rail_fence_decrypt(encrypted, rails)
    
    print(f"Original:  {text}")
    print(f"Encrypted: {encrypted}")
    print(f"Decrypted: {decrypted}")

File: java/Railfence.java
import java.util.Arrays;

public class Railfence {

    public static String encrypt(String text, int rails) {
        if (rails <= 1)
            return text;

        StringBuilder[] fence = new StringBuilder[rails];
        for (int i = 0; i < rails; i++)
            fence[i] = new StringBuilder();

        int currentRail = 0;
        boolean goingDown = false;

        for (char c : text.toCharArray()) {
            fence[currentRail].append(c);

            // Reverse direction if we hit top or bottom
            if (currentRail == 0 || currentRail == rails - 1) {
                goingDown = !goingDown;
            }

            currentRail += goingDown ? 1 : -1;
        }

        StringBuilder result = new StringBuilder();
        for (StringBuilder sb : fence)
            result.append(sb);

        return result.toString();
    }

    public static String decrypt(String text, int rails) {
        if (rails <= 1)
            return text;

        char[][] fence = new char[rails][text.length()];
        for (int i = 0; i < rails; i++)
            Arrays.fill(fence[i], '\n'); // Placeholder

        // Step 1: Mark the pattern with '*'
        boolean goingDown = false;
        int row = 0, col = 0;

        for (int i = 0; i < text.length(); i++) {
            if (row == 0)
                goingDown = true;
            if (row == rails - 1)
                goingDown = false;

            fence[row][col++] = '*';
            row += goingDown ? 1 : -1;
        }

        // Step 2: Fill the fence with the ciphertext characters
        int index = 0;
        for (int i = 0; i < rails; i++) {
            for (int j = 0; j < text.length(); j++) {
                if (fence[i][j] == '*' && index < text.length()) {
                    fence[i][j] = text.charAt(index++);
                }
            }
        }

        // Step 3: Read the zigzag path
        StringBuilder result = new StringBuilder();
        row = 0;
        col = 0;
        goingDown = false;

        for (int i = 0; i < text.length(); i++) {
            if (row == 0)
                goingDown = true;
            if (row == rails - 1)
                goingDown = false;

            if (fence[row][col] != '\n') {
                result.append(fence[row][col]);
                col++;
            }
            row += goingDown ? 1 : -1;
        }

        return result.toString();
    }

    public static void main(String[] args) {
        String text = "HELLO WORLD";
        int rails = 2;

        String encrypted = encrypt(text, rails);
        String decrypted = decrypt(encrypted, rails);

        System.out.println("Original:  " + text);
        System.out.println("Encrypted: " + encrypted);
        System.out.println("Decrypted: " + decrypted);
    }
}
